### PR TITLE
Refactor UserFeed component 

### PR DESCRIPTION
**Pull request by MonkeyDo**
This PR rewrites our UserFeed class component into a functional component, and uses a new /feed POST endpoint to send the required data, as with other endpoints using react-query.

Rewrote the tests as well and we needed a good way to test react-query components. Spent some time to figure everything out but it should server as a good base for other components.

### CODE REVIEW

**Comment by MonkeyDo** [comment]
Deployed on test.LB and got this error trying to load the page:
Gonna look into it tomorrow.

```
2024-04-23 17:03:08,127 listenbrainz.webserver ERROR    Exception on /feed/ [POST]
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/flask/app.py", line 1455, in wsgi_app
    response = self.full_dispatch_request()
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/flask/app.py", line 869, in full_dispatch_request
    rv = self.handle_user_exception(e)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/flask/app.py", line 867, in full_dispatch_request
    rv = self.dispatch_request()
         ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/flask/app.py", line 852, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/flask_login/utils.py", line 290, in decorated_view
    return current_app.ensure_sync(func)(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/code/listenbrainz/listenbrainz/webserver/views/index.py", line 201, in feed
    return jsonify({
           ^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/flask/json/__init__.py", line 170, in jsonify
    return current_app.json.response(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/flask/json/provider.py", line 215, in response
    f"{self.dumps(obj, **dump_args)}\n", mimetype=self.mimetype
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/flask/json/provider.py", line 180, in dumps
    return json.dumps(obj, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/json/__init__.py", line 238, in dumps
    **kw).encode(obj)
          ^^^^^^^^^^^
  File "/usr/local/lib/python3.11/json/encoder.py", line 200, in encode
    chunks = self.iterencode(o, _one_shot=True)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/json/encoder.py", line 258, in iterencode
    return _iterencode(o, 0)
           ^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/flask/json/provider.py", line 120, in _default
    raise TypeError(f"Object of type {type(o).__name__} is not JSON serializable")
TypeError: Object of type RowMapping is not JSON serializable
```

**Comment by MonkeyDo** [comment]
Currently the feed POST endpoint returns all of the events instead of just X recent ones, please wait while I fix this

**Comment by MonkeyDo** [comment]
@anshg1214 The PR is finally ready for review.
I took the opportunity to test out an infinite loading query, now that we have react-query to facilitate this. Took a moment to adapt it and rewrite the tests accordingly, but both the component and the tests should be a good reference for implementation

File: frontend/js/tests/user-feed/UserFeed.test.tsx [hunk 1]
```diff
@@ -19,488 +19,278 @@
...
-        );
-      });
+    await waitFor(() => {
+      // Wait for data to be successfully loaded
+      expect(queryClient.getQueryState(queryKey)?.status).toEqual("success");
     });
+    const timeline = screen.getByTestId("timeline");
+    expect(timeline).toBeInTheDocument();
+    screen.debug(timeline);
+    expect(within(timeline).getAllByRole("listitem")).toHaveLength(7);
```
Comment (anshg1214) [inline frontend/js/tests/user-feed/UserFeed.test.tsx:176]: I believe this was added for testing, and can be removed now

**Review by anshg1214 — CHANGES_REQUESTED** [review]
The PR looks amazing, the queries are much more optimised now. Just a few suggestions.

Once we're done with our refactoring, we can proceed with caching some more queries like Similar Artists

**Comment by MonkeyDo** [comment]
Thanks fir the review, all great feedback :+1: 
I'm really happy I can work with you!

File: .vscode/launch.json [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+                "<node_internals>/**"
+            ],
+            // "program": "${workspaceFolder}/node_modules/.bin/jest",
+            "outFiles": [
+            ],
+            "args": [
+                "frontend/js/tests/user-feed/UserFeed.test.tsx"
+            ],
+             "name": "Debug tests single run",
+
```
Comment (anshg1214) [inline .vscode/launch.json:21]: I believe this file was added while testing. You can safely remove this file.

File: frontend/js/src/user-feed/UserFeed.tsx [hunk 3]
```diff
@@ -622,145 +518,119 @@ export default class UserFeedPage extends React.Component<
...
-                </a>
-              </li>
-            </ul>
-          </div>
-          <div className="col-md-offset-1 col-md-4">
-            <UserSocialNetwork user={currentUser} />
-          </div>
+                  {(isLoading || isFetchingNextPage) && "Loading more..."}
+                  {!isFetchingNextPage &&
+                    (hasNextPage ? "Load More" : "Nothing more to load")}
```
Comment (anshg1214) [inline frontend/js/src/user-feed/UserFeed.tsx:617]: Just realised, the message should not be displayed if it is loading
```suggestion
                  {!(isLoading || isFetchingNextPage) &&
```

**Review by anshg1214 — CHANGES_REQUESTED** [review]
Two final changes, after which it is ready for merge 🚀

**Review by anshg1214 — APPROVED** [review]

**Pull request merged**